Olympus V2 Zap In

Zayen-X opened the pull request:
-needs peer review
-TODO in CheapestBondHelper

FinagleLord committed:
♻️ Initial Commit

-needs peer review
-TODO in CheapestBondHelper

Zayen-X commented on contracts/zaps/OlympusZapV2.sol:
```diff
@@ -0,0 +1,293 @@
+
+    function _exitOlympus(address fromToken, uint256 amount) internal returns (uint256) {
+        if (fromToken == gOHM) {
+            uint256 sOHMRec = IStaking(staking).unwrap(address(this), amount);
```
You don't need to unwrap. the unstake method takes in a `rebase` param that determins if its gohm or not

https://github.com/OlympusDAO/olympus-contracts/blob/main/contracts/Staking.sol#L170-L191

FinagleLord replied:
removed, and fixed params

Zayen-X commented on contracts/zaps/OlympusZapV2.sol:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    /// @notice update state for gOHM
+    function update_approvals(
```
What is this used for?

FinagleLord replied:
Its used for approving the depo for new principals, I also added an _approvalCheck() before any transferFrom is called from the zap to allow the zap to function in the event that an approval is too low.

NOTE: the comment description has also been fixed.

Zayen-X commented on contracts/zaps/CheapestBondHelper.sol:
```diff
@@ -1,56 +1,71 @@
 
-    //////////////// State ////////////////
+    /// @notice used for access control
+    address public olympusDAO = msg.sender;
```
whats the diff between this and setting it in the constructor?

FinagleLord replied:
there isn't one, but for the sake of tidiness I moved it to the constructor 🙂

Zayen-X commented on contracts/zaps/OlympusZapV2.sol:
```diff
@@ -0,0 +1,293 @@
+        if (toToken == gOHM) {
+            IStaking(staking).stake(address(this), amount, false, false);
+            IStaking(staking).claim(address(this), false);
+            uint256 gOHMRec = IStaking(staking).wrap(msg.sender, amount);
```
same thing here. don't need to wrap. stake takes in a `rebase` boolean:
https://github.com/OlympusDAO/olympus-contracts/blob/main/contracts/Staking.sol#L252-L264

FinagleLord replied:
removed, and fixed params

Zayen-X commented on contracts/zaps/OlympusZapV2.sol:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    /// @notice update state for sOHM
+    function update_sOHM(address _sOHM) external onlyOlympusDAO {
```
don't think we need token update methods.

FinagleLord replied:
update token functions have been removed, tokens are now immutable

FinagleLord committed:
♻ Resolved TODO

"TODO: Make sure bonds are not sold out"

FinagleLord committed:
♻ Resolved TODO

"Make sure bonds are not sold out"

calusaca committed:
Resolve stack too deep error from zap in

calusaca committed:
Merge branch 'V2-Zap' of https://github.com/OlympusDAO/olympus-external into V2-Zap

FinagleLord committed:
♻ Resolve PR issue: Tokens are now immutable

"don't think we need token update methods."

FinagleLord committed:
♻ Moved "olympusDAO = msg.sender" to constructor

FinagleLord committed:
♻ Resolved several PR issues:

- "What is this used for?" added more commenting
- _approvalCheck() has been added to all logic in which another contract is attempting to transferFrom the zap contract
if the target isn't approved for the needed amount, it will max approve said target so the next user doesn't have to.

calusaca committed:
Compile ready

calusaca committed:
Zap V2 testing file, first commit

calusaca committed:
Fixed Error: invalid ENS name

suhailgme committed:
Generalize BondHelper

- Adds functionality to get both the cheapest and a specific bond ID for principal

suhailgme committed:
Add SPDX license

suhailgme committed:
Bump major version - Adds ETH check

- approveToken functions add address(0) check to avoid calling ERC20 functions in cases where the token is ETH
- Functionality is otherwise unchanged from V2.2

suhailgme committed:
Update stakingAddress

suhailgme committed:
Split bonding and staking

- ZapStake now handles swapping to OHM and staking
- ZapBond now handles Zapping into an LP and acquiring a bond
- Update approval checks to use inherited _approveToken from base

calusaca committed:
Test file updated to latest contract changes

calusaca committed:
ZapStake unit testing

suhailgme committed:
Add bond ID getter

calusaca committed:
Zapbond to receive bondId as param, del bondHelper

calusaca committed:
Unit testing updated not to use bondHelper

calusaca committed:
Zap Bond update

calusaca committed:
Update security unit testing

suhailgme committed:
Compute slippage offchain and pass it in as an arg

suhailgme committed:
Example bond test

suhailgme committed:
Add token->token bond example

calusaca committed:
LP tokens zap bonds unit testing

calusaca committed:
ZapBonds unit testing

calusaca committed:
Test non-approved swap target not allowed to zapIn

suhailgme committed:
Fix UST address

suhailgme committed:
Merge branch 'V2-Zap' of https://github.com/OlympusDAO/olympus-external into V2-Zap

suhailgme committed:
Rename circuit breaker

suhailgme committed:
Cleanup

- Update and fix lots of comments
- Explicit solidity version
- Remove unneeded constructor params
- Pass correct token to zapStake event
- Rename _enterOlympus to _stake to better describe purpose
- Remove unneeded functions

suhailgme committed:
Cleanup and update tests

- Added tests for slippage
- Fix tests for authorized targets
- Add BondId constants
- Move olympus staking to constants

suhailgme committed:
Add pausable test for ZapBond

Zayen-X commented on contracts/zaps/OlympusZapV2.sol:
```diff
@@ -0,0 +1,198 @@
+
+        if (toToken == gOHM) {
+            // stake OHM -> gOHM
+            IStaking(staking).stake(address(this), amount, false, false);
```
The last param of this method is `_claim` which removes the need for the additional call in the next line. You can do this for the sOHM call as well.

Zayen-X commented on contracts/zaps/OlympusZapV2.sol:
```diff
@@ -0,0 +1,198 @@
+        address toToken,
+        uint256 minToToken,
+        address swapTarget,
+        bytes calldata swapData,
```
Where are docs/codes on how we're generating this?

Zayen-X commented on contracts/zaps/OlympusZapV2.sol:
```diff
@@ -0,0 +1,198 @@
+        uint256 minToToken,
+        address swapTarget,
+        bytes calldata swapData,
+        address referral
```
Whats this used for?

Zayen-X commented on test/zaps/OlympusZapV2.test.ts:
```diff
@@ -0,0 +1,572 @@
+
+const OlympusZapArtifact = "Olympus_V2_Zap_In";
+
+describe("OlympusDAO Zap", () => {
```
Thank you so much for putting this together. 🙏

Zayen-X commented on libs/exchange/exchange.helper.ts:
```diff
@@ -30,3 +30,28 @@ export const exchangeAndApprove = async (
   return await token.balanceOf(await fromAccount.getAddress());
 };
+
+export const exchange = async (
```
I think the file name is weird.

Zayen-X reviewed (commented):
Initial pass at solidity. Haven't looked at tests.

suhailgme committed:
Use claim flag instead of extra call to claim fx

suhailgme committed:
Remove olympusDAO address

- Contract will be owned by olympusDAO as owner() so this extra state is not needed
- Update tests for change

suhailgme committed:
Remove .only

calusaca committed:
Deployment unit testing

suhailgme committed:
Update OlympusZapV2.sol

suhailgme committed:
Merge branch 'V2-Zap' of https://github.com/OlympusDAO/olympus-external into V2-Zap

suhailgme committed:
Revert cb2110e

suhailgme committed:
Setup deployment pattern

- Create deploy script
- Update hardhat.config for mnemonic

suhailgme committed:
Set approved targets as needed by gov

suhailgme committed:
Deploy and verify

- Deployed at https://etherscan.io/address/0xc1faf9ad8a5a7d0c328f1f395a150fab293e8bbd#code
